File: 2021/03/03b.swift
func boolArray2Int(boolArray: [Int]) -> Int {
  let nBits = boolArray.count
  var result = 0
  for i in 0..<nBits {
    if boolArray[i] == 1 {
      result += 1 << (nBits - i - 1)
    }
  }
  return result
}

func oneIsMost(input: [Int], def: Bool, anti: Bool) -> Bool {
  let sum = input.reduce(0) { $0 + $1 }
  if Double(sum) == Double(input.count) / 2.0 {
    return def
  }
  return (sum > input.count / 2) != anti
}

func rating(lines: [[Int]], def: Bool, anti: Bool) -> [Int] {
  let nBits = lines[0].count
  var results = lines
  for i in 0..<nBits {
    let transposed = results.map { $0[i] }
    let modeAtI = oneIsMost(input: transposed, def: def, anti: anti)
    results = results.filter { ($0[i] == 1) == modeAtI }
    if results.count <= 1 {
      break
    }
  }
  return results[0]
}

var lines: [[Int]] = []
while let line = readLine() {
  lines.append(Array(line).map { $0.wholeNumberValue! })
}

let ox = rating(lines: lines, def: true, anti: false)
let co2 = rating(lines: lines, def: false, anti: true)
print(boolArray2Int(boolArray: ox) * boolArray2Int(boolArray: co2))

File: 2021/03/03b.py
from sys import stdin

def mode(nums, default: str, anti):
    if nums.count("1") == nums.count("0"):
        return default
    return "0" if (nums.count("1") <= len(nums) // 2) != anti else "1"

def rating(lines, default: str, anti = False):
    results = lines
    for i in range(len(lines[0])):
        mode_at_i = mode(list(zip(*results))[i], default, anti)
        results = [line for line in results if line[i] == mode_at_i]
        if len(results) <= 1:
            break
    return results

lines = [line.strip() for line in stdin.readlines()]
ox = rating(lines, "1")
co2 = rating(lines, "0", anti=True)
print(int("".join(ox), 2) * int("".join(co2), 2))
